## Final Project Submission

Please fill out:
* Student name: Daisy Wagati 
* Student pace: Full Time 
* Instructor name: Mwikali 


## Project Overview

This project deals with a company that is seeking to expand to new industries to expand its portfolio by  purchasing and operating airplanes for commercial and private enterprises. Through a comprehensive analysis of aviation accident data from 1962 to 2023, we will conduct an exploratory data analysis to provide insights on the frequency and distribution of accidents over time and across different airports. Ultimately, our goal is to provide actionable insights that will guide the company's entry into the aircraft industry, and set it up for long-term success and profitability.

### Import Relevant Libraries to Notebook

In the cell below, we will import the relevant libraries that will enable us to read data from the csv file.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

### Load the dataset into a dataframe

In the cell below, we will use pandas to create a new DataFrame called **df** containing the data from the dataset in the file **Aviation_Data.csv** in the folder containing this notebook. We will then check the structure of our data by calling **df.head()** which shows us the preview of our data


In [2]:
#Use pandas to load the csv file
df = pd.read_csv('data/Aviation_Data.csv', low_memory=False)
df.head()

Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country   Latitude   Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
0          NaN  ...          Personal         NaN                  2.0   
1          NaN  ...          Personal         NaN                  4.0   
2          NaN  ...          Personal         NaN                  3.0   
3          NaN  ...          Personal         NaN                  2.0   
4          NaN  ...          Personal         NaN                  1.0   

  Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
0                    0.0                  0.0             0.0   
1                    0.0                  0.0             0.0   
2                    NaN                  NaN             NaN   
3                    0.0                  0.0             0.0   
4                    2.0                  NaN             0.0   

  Weather.Condition  Broad.phase.of.flight   Report.Status Publication.Date  
0               UNK                 Cruise  Probable Cause              NaN  
1               UNK                Unknown  Probable Cause       19-09-1996  
2               IMC                 Cruise  Probable Cause       26-02-2007  
3               IMC                 Cruise  Probable Cause       12-09-2000  
4               VMC               Approach  Probable Cause       16-04-1980  

[5 rows x 31 columns]

### Data Understanding

Here, we would like to get familiar with the data by understanding the dimensionality of our dataset and investigating what type of data it contains, and the data types used to store it

We will start by calling **.shape** to get the number of rows and columns in the DataFrame

In [3]:
#Use .shape to get number of rows and columns
df.shape

(90348, 31)

Next we call **.info()** to get a summary of the DataFrame's structure and content

In [4]:
#Use .info() to get summary of DataFrame's structure and content
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90348 entries, 0 to 90347
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      90348 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50249 non-null  object 
 9   Airport.Name            52790 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87572 non-null  object 
 14  Make                    88826 non-null

Then we call **.describe()** to get descriptive statistics of our DataFrame which gives us a summary of central tendency, dispersion and shape of a data set's distribution

In [5]:
df.describe()

Number.of.Engines  Total.Fatal.Injuries  Total.Serious.Injuries  \
count       82805.000000          77488.000000            76379.000000   
mean            1.146585              0.647855                0.279881   
std             0.446510              5.485960                1.544084   
min             0.000000              0.000000                0.000000   
25%             1.000000              0.000000                0.000000   
50%             1.000000              0.000000                0.000000   
75%             1.000000              0.000000                0.000000   
max             8.000000            349.000000              161.000000   

       Total.Minor.Injuries  Total.Uninjured  
count          76956.000000     82977.000000  
mean               0.357061         5.325440  
std                2.235625        27.913634  
min                0.000000         0.000000  
25%                0.000000         0.000000  
50%                0.000000         1.000000  
75%                0.000000         2.000000  
max              380.000000       699.000000

### Data Cleaning

We shall perform data cleaning so as to detect and correct(or remove) corrupt or inaccurate records from the data set, and improve overall quality and readability of the data set. We shall handle missing values, outliers and duplicates, as well as transforming and formatting the data into a consistent and usable format.

We shall start by creating a copy of the original DataFrame before we make any changes to it in case we would like to retrieve original information later.

In [6]:
#Create copy of the DataFrame
df_copy = df.copy

### Drop unnecessary rows

We shall then look for and drop duplicates so as to get rid of unnecessary rows

In [7]:
#Identify duplicates
df.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
90343    False
90344    False
90345    False
90346    False
90347    False
Length: 90348, dtype: bool

In [8]:
#Check sum of duplicates
df.duplicated().sum()

1390

In [9]:
#Drop duplicates
df.drop_duplicates(inplace = True, keep = 'first')

In [10]:
#Confirm that duplicates have been dropped by calling **.shape**
df.shape

(88958, 31)

As can be seen above duplicates have been dropped, reducing the number of rows from 90348 to 88958

### Drop unnecessary Columns
We have columns that are unnecessary to our project and we need to drop them.

In [11]:
#Drop unnecessary columns
df.drop(columns=['Event.Id', 'Accident.Number', 'Location', 'Latitude','Longitude','Airport.Code', 'FAR.Description', 'Schedule'], inplace = True)

In [12]:
#Confirm that unnecessary columns have been dropped
df.shape

(88958, 23)

As can be seen above, unnecessary columns have been dropped reducing number of columns from 31 to 23. Now we can work with the narrowed down data.

In [13]:
#Check for missing values
df.isna()

Investigation.Type  Event.Date  Country  Airport.Name  Injury.Severity  \
0                   False       False    False          True            False   
1                   False       False    False          True            False   
2                   False       False    False          True            False   
3                   False       False    False          True            False   
4                   False       False    False          True            False   
...                   ...         ...      ...           ...              ...   
90343               False       False    False          True            False   
90344               False       False    False          True             True   
90345               False       False    False         False            False   
90346               False       False    False          True             True   
90347               False       False    False          True            False   

       Aircraft.damage  Aircraft.Category  Registration.Number   Make  Model  \
0                False               True                False  False  False   
1                False               True                False  False  False   
2                False               True                False  False  False   
3                False               True                False  False  False   
4                False               True                False  False  False   
...                ...                ...                  ...    ...    ...   
90343             True               True                False  False  False   
90344             True               True                False  False  False   
90345            False              False                False  False  False   
90346             True               True                False  False  False   
90347             True               True                False  False  False   

       ...  Purpose.of.flight  Air.carrier  Total.Fatal.Injuries  \
0      ...              False         True                 False   
1      ...              False         True                 False   
2      ...              False         True                 False   
3      ...              False         True                 False   
4      ...              False         True                 False   
...    ...                ...          ...                   ...   
90343  ...              False         True                 False   
90344  ...               True         True                 False   
90345  ...              False         True                 False   
90346  ...              False        False                 False   
90347  ...              False         True                 False   

       Total.Serious.Injuries  Total.Minor.Injuries  Total.Uninjured  \
0                       False                 False            False   
1                       False                 False            False   
2                        True                  True             True   
3                       False                 False            False   
4                       False                  True            False   
...                       ...                   ...              ...   
90343                   False                 False            False   
90344                   False                 False            False   
90345                   False                 False            False   
90346                   False                 False            False   
90347                   False                 False            False   

       Weather.Condition  Broad.phase.of.flight  Report.Status  \
0                  False                  False          False   
1                  False                  False          False   
2                  False                  False          False   
3                  False                  False          False   
4                  False                  False          Fals

In [14]:
#Check sum of missing values 
df.isna().sum()

Investigation.Type            0
Event.Date                   69
Country                     295
Airport.Name              36168
Injury.Severity            1069
Aircraft.damage            3263
Aircraft.Category         56671
Registration.Number        1386
Make                        132
Model                       161
Amateur.Built               171
Number.of.Engines          6153
Engine.Type                7146
Purpose.of.flight          6261
Air.carrier               72310
Total.Fatal.Injuries      11470
Total.Serious.Injuries    12579
Total.Minor.Injuries      12002
Total.Uninjured            5981
Weather.Condition          4561
Broad.phase.of.flight     27234
Report.Status              6450
Publication.Date          15299
dtype: int64

In [15]:
#Replace missing numeric values with median
df.fillna(df.median(), inplace = True)
df.head()

Investigation.Type  Event.Date        Country Airport.Name Injury.Severity  \
0           Accident  1948-10-24  United States          NaN        Fatal(2)   
1           Accident  1962-07-19  United States          NaN        Fatal(4)   
2           Accident  1974-08-30  United States          NaN        Fatal(3)   
3           Accident  1977-06-19  United States          NaN        Fatal(2)   
4           Accident  1979-08-02  United States          NaN        Fatal(1)   

  Aircraft.damage Aircraft.Category Registration.Number      Make     Model  \
0       Destroyed               NaN              NC6404   Stinson     108-3   
1       Destroyed               NaN              N5069P     Piper  PA24-180   
2       Destroyed               NaN              N5142R    Cessna      172M   
3       Destroyed               NaN              N1168J  Rockwell       112   
4       Destroyed               NaN               N15NY    Cessna       501   

   ... Purpose.of.flight  Air.carrier Total.Fatal.Injuries  \
0  ...          Personal          NaN                  2.0   
1  ...          Personal          NaN                  4.0   
2  ...          Personal          NaN                  3.0   
3  ...          Personal          NaN                  2.0   
4  ...          Personal          NaN                  1.0   

  Total.Serious.Injuries Total.Minor.Injuries  Total.Uninjured  \
0                    0.0                  0.0              0.0   
1                    0.0                  0.0              0.0   
2                    0.0                  0.0              1.0   
3                    0.0                  0.0              0.0   
4                    2.0                  0.0              0.0   

   Weather.Condition  Broad.phase.of.flight   Report.Status Publication.Date  
0                UNK                 Cruise  Probable Cause              NaN  
1                UNK                Unknown  Probable Cause       19-09-1996  
2                IMC                 Cruise  Probable Cause       26-02-2007  
3                IMC                 Cruise  Probable Cause       12-09-2000  
4                VMC               Approach  Probable Cause       16-04-1980  

[5 rows x 23 columns]

In [16]:
df.tail()

Investigation.Type  Event.Date        Country Airport.Name  \
90343           Accident  2022-12-26  United States          NaN   
90344           Accident  2022-12-26  United States          NaN   
90345           Accident  2022-12-26  United States       PAYSON   
90346           Accident  2022-12-26  United States          NaN   
90347           Accident  2022-12-29  United States          NaN   

      Injury.Severity Aircraft.damage Aircraft.Category Registration.Number  \
90343           Minor             NaN               NaN              N1867H   
90344             NaN             NaN               NaN              N2895Z   
90345       Non-Fatal     Substantial          Airplane              N749PJ   
90346             NaN             NaN               NaN              N210CU   
90347           Minor             NaN               NaN              N9026P   

                             Make      Model  ... Purpose.of.flight  \
90343                       PIPER  PA-28-151  ...          Personal   
90344                    BELLANCA       7ECA  ...               NaN   
90345  AMERICAN CHAMPION AIRCRAFT      8GCBC  ...          Personal   
90346                      CESSNA       210N  ...          Personal   
90347                       PIPER  PA-24-260  ...          Personal   

              Air.carrier Total.Fatal.Injuries Total.Serious.Injuries  \
90343                 NaN                  0.0                    1.0   
90344                 NaN                  0.0                    0.0   
90345                 NaN                  0.0                    0.0   
90346  MC CESSNA 210N LLC                  0.0                    0.0   
90347                 NaN                  0.0                    1.0   

      Total.Minor.Injuries  Total.Uninjured  Weather.Condition  \
90343                  0.0              0.0                NaN   
90344                  0.0              0.0                NaN   
90345                  0.0              1.0                VMC   
90346                  0.0              0.0                NaN   
90347                  0.0              1.0                NaN   

       Broad.phase.of.flight  Report.Status Publication.Date  
90343                    NaN            NaN       29-12-2022  
90344                    NaN            NaN              NaN  
90345                    NaN            NaN       27-12-2022  
90346                    NaN            NaN              NaN  
90347                    NaN            NaN       30-12-2022  

[5 rows x 23 columns]

In [17]:
#Replacing NaN Categorical Values with a Placeholder
df.fillna('?', inplace = True)
df.head()


Investigation.Type  Event.Date        Country Airport.Name Injury.Severity  \
0           Accident  1948-10-24  United States            ?        Fatal(2)   
1           Accident  1962-07-19  United States            ?        Fatal(4)   
2           Accident  1974-08-30  United States            ?        Fatal(3)   
3           Accident  1977-06-19  United States            ?        Fatal(2)   
4           Accident  1979-08-02  United States            ?        Fatal(1)   

  Aircraft.damage Aircraft.Category Registration.Number      Make     Model  \
0       Destroyed                 ?              NC6404   Stinson     108-3   
1       Destroyed                 ?              N5069P     Piper  PA24-180   
2       Destroyed                 ?              N5142R    Cessna      172M   
3       Destroyed                 ?              N1168J  Rockwell       112   
4       Destroyed                 ?               N15NY    Cessna       501   

   ... Purpose.of.flight  Air.carrier Total.Fatal.Injuries  \
0  ...          Personal            ?                  2.0   
1  ...          Personal            ?                  4.0   
2  ...          Personal            ?                  3.0   
3  ...          Personal            ?                  2.0   
4  ...          Personal            ?                  1.0   

  Total.Serious.Injuries Total.Minor.Injuries  Total.Uninjured  \
0                    0.0                  0.0              0.0   
1                    0.0                  0.0              0.0   
2                    0.0                  0.0              1.0   
3                    0.0                  0.0              0.0   
4                    2.0                  0.0              0.0   

   Weather.Condition  Broad.phase.of.flight   Report.Status Publication.Date  
0                UNK                 Cruise  Probable Cause                ?  
1                UNK                Unknown  Probable Cause       19-09-1996  
2                IMC                 Cruise  Probable Cause       26-02-2007  
3                IMC                 Cruise  Probable Cause       12-09-2000  
4                VMC               Approach  Probable Cause       16-04-1980  

[5 rows x 23 columns]

In [18]:
#Check for unique values in the dataset
df['Aircraft.Category'].unique()

array(['?', 'Airplane', 'Helicopter', 'Glider', 'Balloon', 'Gyrocraft',
       'Ultralight', 'Unknown', 'Blimp', 'Powered-Lift', 'Weight-Shift',
       'Powered Parachute', 'Rocket', 'WSFT', 'UNK', 'ULTR'], dtype=object)

In [19]:
#Check for how many times each unique value is present
df['Aircraft.Category'].value_counts()

?                    56671
Airplane             27617
Helicopter            3440
Glider                 508
Balloon                231
Gyrocraft              173
Weight-Shift           161
Powered Parachute       91
Ultralight              30
Unknown                 14
WSFT                     9
Powered-Lift             5
Blimp                    4
UNK                      2
ULTR                     1
Rocket                   1
Name: Aircraft.Category, dtype: int64

In [20]:
#Confirm that there are no missing values and that our data is clean
df.isna().sum()

Investigation.Type        0
Event.Date                0
Country                   0
Airport.Name              0
Injury.Severity           0
Aircraft.damage           0
Aircraft.Category         0
Registration.Number       0
Make                      0
Model                     0
Amateur.Built             0
Number.of.Engines         0
Engine.Type               0
Purpose.of.flight         0
Air.carrier               0
Total.Fatal.Injuries      0
Total.Serious.Injuries    0
Total.Minor.Injuries      0
Total.Uninjured           0
Weather.Condition         0
Broad.phase.of.flight     0
Report.Status             0
Publication.Date          0
dtype: int64